### ISSUE TITLE
Helm/--set replaces entire sequence when modifying sub-sequence

### ISSUE THREAD

**Issue by jasondemps1** [helm/helm#8231]
Output of `helm version`: v3.2.0

Output of `kubectl version`: N/A

Cloud Provider/Platform (AKS, GKE, Minikube etc.):  N/A

I'm working through a basic problem with YAML sequences and trying to override specific positions with --set. Based on the Helm3 docs, if this is my values.yaml:

```
servers:
  - name: foo
    port: 80
  - name: bar
    port: 81
```

the way the documentation reads, I should be able to modify one of the ports via:

`--set servers[0].port=value`

The issue I'm running into is this doesn't seem to be the case, when I execute this via:

`helm template --dry-run --debug --generate-name chart-path --set servers[0].port=123`

the result is:

```
# Source: test/templates/services.yaml
name:
port: 123
```

It seems to destroy/modify the entire sequence.

What I am expecting to see is:

```
# Source: test/templates/services.yaml
name: foo
port: 123
name: bar
port: 443
```

In other words, I'm trying to modify a value within the sequence in-place and not modify or touch other positions. Maybe I'm just really misunderstanding? Is there a way to achieve what I'm looking for if this isn't the solution? Thank you!

Doc reference: https://helm.sh/docs/chart_best_practices/values/ (Section: **Consider how users will use your values**)

**Comment by bacongobbler** [comment]
The syntax you describe does not replace the port for the server at index 0. it's the equivalent of

```
servers:
  - port: 123
```

https://helm.sh/docs/intro/using_helm/#the-format-and-limitations-of---set

Consider using a dictionary with the name as a unique index instead:

```
servers:
  foo:
    port: 80
  bar:
    port: 81
```

Then you can change port numbers with `--set servers.foo.port=123`.

Hope this helps.
